patches/tModLoader/Terraria/ModLoader/Config/ConfigManager.cs

```diff
@@ -72,6 +74,113 @@ internal static void Add(ModConfig config)
+				if (variable.Type.IsClass && variable.Type.Assembly == type.Assembly) {
+#pragma warning disable CS0618 // Type or member is obsolete
+					var typeLabelObsolete = (LabelAttribute)Attribute.GetCustomAttribute(variable.Type, typeof(LabelAttribute));
+#pragma warning restore CS0618 // Type or member is obsolete
+
+					var typeLabel = (LabelKeyAttribute)Attribute.GetCustomAttribute(variable.Type, typeof(LabelKeyAttribute));
+					if (typeLabel != null) {
+						if (typeLabel.malformed)
+							throw new ValueNotTranslationKeyException($"{nameof(LabelKeyAttribute)} only accepts localization keys for the 'key' parameter.", errorOnType: true);
+					}
+					else { 
+						string typeLabelKey = config.Mod.GetLocalizationKey($"Configs.{variable.Type.Name}.Label");
+						string typeLabelValue = Regex.Replace(variable.Type.Name, "([A-Z])", " $1").Trim();
+						if (typeLabelObsolete != null)
+							typeLabelValue = typeLabelObsolete.LocalizationEntry;
+						Language.GetOrRegister(typeLabelKey, () => typeLabelValue);
+					}
+
+					var typeTooltip = (TooltipKeyAttribute)Attribute.GetCustomAttribute(variable.Type, typeof(TooltipKeyAttribute));
+					if (typeTooltip != null) {
+						if (typeTooltip.malformed)
+							throw new ValueNotTranslationKeyException($"{nameof(TooltipKeyAttribute)} only accepts localization keys for the 'key' parameter.", errorOnType: true);
+					}
+					else {
+						string typeTooltipKey = config.Mod.GetLocalizationKey($"Configs.{variable.Type.Name}.Tooltip");
+						string typeTooltipValue = "";
+						Language.GetOrRegister(typeTooltipKey, () => typeTooltipValue);
+					}
+					RegisterLocalizationKeysForMembers(config, variable.Type);
+				}
+
+				// Handle obsolete attributes. Use them to populate value of key, if present, to ease porting.
+#pragma warning disable CS0618 // Type or member is obsolete
+				var labelObsolete = (LabelAttribute)Attribute.GetCustomAttribute(variable.MemberInfo, typeof(LabelAttribute));
+				var tooltipObsolete = (TooltipAttribute)Attribute.GetCustomAttribute(variable.MemberInfo, typeof(TooltipAttribute));
+#pragma warning restore CS0618 // Type or member is obsolete
+
+				// Label and Tooltip will always exist. Header is optional, need to be used to exist.
+				var header = GetLocalizedHeader(variable, config, throwErrors: true);
+				if (header != null) {
+					string identifier = header.IsIdentifier ? header.identifier : variable.Name;
+					Language.GetOrRegister(header.key, () => $"{Regex.Replace(identifier, "([A-Z])", " $1").Trim()} Header");
+				}
+
+				string labelKey = GetLabelKey(variable, config, fallbackToClass: false, throwErrors: true);
+				if (labelKey.StartsWith($"Mods.{config.Mod.Name}")) {
+					string labelValue = Regex.Replace(variable.Name, "([A-Z])", " $1").Trim();
+					if (labelObsolete != null)
+						labelValue = labelObsolete.LocalizationEntry;
```

Comment:
null coalescing? put inside the `() => ` so it only runs when necessary. Not actually harmful for performance, but is the intended pattern